# Regras de associação.

Aluno: **Pedro Henrique Rodrigues de Lima**


## Introdução
### Regras de associação geradas a partir de itemsets frequentes

O objetivo dessa atividade é encontrar pessoas em organizações comunitárias de uma cidade que possuam algum tipo de conexão com as outras, para que elas possam, no futuro, trabalharem de forma cooperativa.

Para alcançar este objetivo, iremos utilizar o modelo de regras de associação. Então nosso objetivo torna-se encontrar regras de associação que modelem ligações entre tipos de organizações comunitárias.
Resumo dos dados:

  * Familiar: Sim/Não indicando se é membro de uma organização orientada à família como grupos de apadrinhamento.
   * Hobbies: Sim/Não indicando se é membro de uma organização orientada hobby como rádio amador, aeromodelismo, motocicleta, etc.
   * Clubes: Sim/Não indicando se é membro de uma organização social como o Lions Club ou Rotary Club, etc.
   * Política: Sim/Não indicando se é membro de uma organização política como partidos políticos, grupos de lobby, etc.
   * Profissional: Sim/Não indicando se é membro de uma organização profissional como a OAB, o CRM, CRA, etc.
   * Religiosa: Sim/Não indicando se é membro de uma organização religiosa como uma igreja.
   * Grupo de suporte: Sim/Não indicando se é membro de uma organização de suporte à comunidade, como os Alcóolicos Anônimos, LBV, Médicos sem Fronteiras, etc.



## Leitura dos dados e importação das bibliotecas


In [25]:
#! pip install mlxtend
#! pip install xlrd

import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [35]:

dataset = pd.read_excel('Atividade - Regras de Associação - Comunidade.xlsx')

#seleciona apenas as colunas com os tipos de grupo frequentado
dataset = dataset.iloc[:,5:]
dataset = dataset.replace({1:True,0:False})
dataset


Familiar  Hobbies  Clubes  Política  Profissional  Religiosa  \
0         True    False   False     False         False      False   
1        False    False   False     False         False       True   
2         True     True   False     False          True      False   
3        False    False   False     False         False      False   
4        False    False   False      True          True      False   
...        ...      ...     ...       ...           ...        ...   
3478     False    False   False     False         False      False   
3479     False    False   False     False         False       True   
3480     False     True   False     False         False       True   
3481      True     True    True     False          True       True   
3482      True    False   False     False          True      False   

      Grupo de suporte  
0                False  
1                 True  
2                False  
3                False  
4                 True  
...                ...  
3478             False  
3479             False  
3480              True  
3481             False  
3482             False  

[3483 rows x 7 columns]

## Calculo dos itensets frequentes e regras de associação


**Primeira etapa: Definição dos itensets frequentes:**

**Itensets frequentes com suporte mínimo de 0.15**:

In [30]:
frequent_itemsets = apriori(dataset, min_support=0.15, use_colnames=True)
print(frequent_itemsets)

   support                        itemsets
0     0.39                      (Familiar)
1     0.30                       (Hobbies)
2     0.19                        (Clubes)
3     0.32                  (Profissional)
4     0.42                     (Religiosa)
5     0.16              (Grupo de suporte)
6     0.19             (Hobbies, Familiar)
7     0.22           (Familiar, Religiosa)
8     0.24            (Hobbies, Religiosa)
9     0.15  (Hobbies, Familiar, Religiosa)


**Segunda etapa: "Cálculo" das regras de associação de acordo com alguma medida de associação:**


**Regras de associação com confiança mínima de 0.5:**

In [32]:
association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5)

antecedents            consequents  antecedent support  \
0              (Hobbies)             (Familiar)                0.30   
1             (Familiar)            (Religiosa)                0.39   
2            (Religiosa)             (Familiar)                0.42   
3              (Hobbies)            (Religiosa)                0.30   
4            (Religiosa)              (Hobbies)                0.42   
5    (Hobbies, Familiar)            (Religiosa)                0.19   
6   (Hobbies, Religiosa)             (Familiar)                0.24   
7  (Familiar, Religiosa)              (Hobbies)                0.22   
8              (Hobbies)  (Familiar, Religiosa)                0.30   

   consequent support  support  confidence  lift  leverage  conviction  
0                0.39     0.19        0.62  1.60      0.07        1.62  
1                0.42     0.22        0.58  1.38      0.06        1.37  
2                0.39     0.22        0.54  1.38      0.06        1.32  
3                0.42     0.24        0.80  1.90      0.11        2.85  
4                0.30     0.24        0.57  1.90      0.11        1.63  
5                0.42     0.15        0.83  1.98      0.08        3.38  
6                0.39     0.15        0.65  1.66      0.06        1.73  
7                0.30     0.15        0.69  2.30      0.09        2.25  
8                0.22     0.15        0.52  2.30      0.09        1.60

**Regras de associação com lift mínimo de 1.3:**

In [33]:
association_rules(frequent_itemsets, metric="lift", min_threshold=1.3)

antecedents            consequents  antecedent support  \
0               (Hobbies)             (Familiar)                0.30   
1              (Familiar)              (Hobbies)                0.39   
2              (Familiar)            (Religiosa)                0.39   
3             (Religiosa)             (Familiar)                0.42   
4               (Hobbies)            (Religiosa)                0.30   
5             (Religiosa)              (Hobbies)                0.42   
6     (Hobbies, Familiar)            (Religiosa)                0.19   
7    (Hobbies, Religiosa)             (Familiar)                0.24   
8   (Familiar, Religiosa)              (Hobbies)                0.22   
9               (Hobbies)  (Familiar, Religiosa)                0.30   
10             (Familiar)   (Hobbies, Religiosa)                0.39   
11            (Religiosa)    (Hobbies, Familiar)                0.42   

    consequent support  support  confidence  lift  leverage  conviction  
0                 0.39     0.19        0.62  1.60      0.07        1.62  
1                 0.30     0.19        0.48  1.60      0.07        1.34  
2                 0.42     0.22        0.58  1.38      0.06        1.37  
3                 0.39     0.22        0.54  1.38      0.06        1.32  
4                 0.42     0.24        0.80  1.90      0.11        2.85  
5                 0.30     0.24        0.57  1.90      0.11        1.63  
6                 0.42     0.15        0.83  1.98      0.08        3.38  
7                 0.39     0.15        0.65  1.66      0.06        1.73  
8                 0.30     0.15        0.69  2.30      0.09        2.25  
9                 0.22     0.15        0.52  2.30      0.09        1.60  
10                0.24     0.15        0.40  1.66      0.06        1.26  
11                0.19     0.15        0.37  1.98      0.08        1.29

## Comentários

1. A quantidade de indivíduos participando de grupos políticos é pequena, e esse tipo de grupo não apareceu nem mesmo no filtro de suporte de 0.15.


2. O dataset é pequeno (3483), mas conseguimos enxergar algumas regras fortes entre os grupos Hobbies, Familiar e Religiosa. Pessoas frequentando algum desses grupos tem uma grande probabilidade de participar de outro grupo, tendo em vista o lift bem superior a 1 e o suporte maior do que 0.5.